src/core/ext/filters/client_channel/lb_policy/xds/xds_load_balancer_api.cc

```diff
@@ -18,62 +18,167 @@
+      request, upb_strview_make(kEdsTypeUrl, sizeof(kEdsTypeUrl) - 1));
+  // Encode the request.
+  size_t output_length;
+  char* output = envoy_api_v2_DiscoveryRequest_serialize(request, arena.ptr(),
+                                                         &output_length);
+  return grpc_slice_from_copied_buffer(output, output_length);
+}
+
+namespace {
+
+grpc_error* ServerAddressParseAndAppend(const XdsLbEndpoint* lb_endpoint,
+                                        ServerAddressList* list) {
+  // Find the ip:port.
+  const XdsEndpoint* endpoint =
+      envoy_api_v2_endpoint_LbEndpoint_endpoint(lb_endpoint);
+  const XdsAddress* address = envoy_api_v2_endpoint_Endpoint_address(endpoint);
+  const XdsSocketAddress* socket_address =
+      envoy_api_v2_core_Address_socket_address(address);
+  upb_strview address_strview =
+      envoy_api_v2_core_SocketAddress_address(socket_address);
+  uint32_t port = envoy_api_v2_core_SocketAddress_port_value(socket_address);
+  if (GPR_UNLIKELY(port >> 16) != 0) {
+    return GRPC_ERROR_CREATE_FROM_STATIC_STRING("Invalid port.");
   }
-  ++sl->num_servers;
-  return true;
-}
-
-typedef struct decode_serverlist_arg {
-  /* The decoding callback is invoked once per server in serverlist. Remember
-   * which index of the serverlist are we currently decoding */
-  size_t decoding_idx;
-  /* The decoded serverlist */
-  xds_grpclb_serverlist* serverlist;
-} decode_serverlist_arg;
-
-/* invoked once for every Server in ServerList */
-static bool decode_serverlist(pb_istream_t* stream, const pb_field_t* field,
-                              void** arg) {
-  decode_serverlist_arg* dec_arg = static_cast<decode_serverlist_arg*>(*arg);
-  GPR_ASSERT(dec_arg->serverlist->num_servers >= dec_arg->decoding_idx);
-  xds_grpclb_server* server =
-      static_cast<xds_grpclb_server*>(gpr_zalloc(sizeof(xds_grpclb_server)));
-  if (GPR_UNLIKELY(!pb_decode(stream, grpc_lb_v1_Server_fields, server))) {
-    gpr_free(server);
-    gpr_log(GPR_ERROR, "nanopb error: %s", PB_GET_ERROR(stream));
-    return false;
+  // Populate grpc_resolved_address.
+  grpc_resolved_address addr;
+  char* address_str = static_cast<char*>(gpr_malloc(address_strview.size + 1));
+  memcpy(address_str, address_strview.data, address_strview.size);
+  address_str[address_strview.size] = '\0';
+  grpc_string_to_sockaddr(&addr, address_str, port);
+  gpr_free(address_str);
+  // Append the address to the list.
+  list->emplace_back(addr, nullptr);
+  return GRPC_ERROR_NONE;
+}
+
+grpc_error* LocalityParse(const XdsLocalityLbEndpoints* locality_lb_endpoints,
+                          XdsLocalityInfo* locality_info) {
+  upb::Arena arena;
+  // Parse locality name.
+  const XdsLocality* locality =
+      envoy_api_v2_endpoint_LocalityLbEndpoints_locality(locality_lb_endpoints);
+  locality_info->locality_name = MakeRefCounted<XdsLocalityName>(
+      UniquePtr<char>(
+          gpr_strdup(envoy_api_v2_core_Locality_region(locality).data)),
```

Comment:
Is this relying on the string to be NULL-terminated? That is not a safe assumption -- upb will make this a pointer into the input buffer you parsed from, and the string member will not have a NULL terminator.